Wednesday Homework

jblneely opened the pull request:
Please describe your comfort and completeness levels before submitting.

Comfort Level (1-5):
3
Completeness Level (1-5):
4
What did you think of this deliverable?
useful skills for working with functions: adding and amending properties to objects

jblneely committed:
Wednesday Homework

connorjclark commented:
Bowman

## Syntax Errors

Use the Chrome Console to view where the syntax errors in your code are. Most of them are missing commas, one of them is a missing closing brace "}".

## `animals[i].makeNoise(this)`

Your animal objects have a `makeNoise` function. `makeNoise` doesn't have any parameters, but here you are calling `makeNoise` with a `this`.

You can just do `animals[i].makeNoise()`. Furthermore, `this` only exists within the context of an object (or a function), but this code is just in a for-loop.


## makeFriend

```
var makeFriend = function(newFriend) {
    // your code here
    // hint: use the 'this' keyword
};

duck.makeFriend = makeFriend;
swan.makeFriend = makeFriend;
```

Here, we are adding the same function to `duck` and `swan`. This function takes one parameter, `newFriend`. This function should add that `newFriend` to the object's `friends` array.

If this is the same function, how can we make sure we are referencing the correct `friends`? The `this` variable refers to the object that called the function. For example,

```
myObject.myFunction();
```

Here, our code inside `myFunction` will have access to a variable called `this`, which will refer to `myObject`.

So,

```
duck.makeFriend(swan);
```

Here, `makeFriend` has access to a variable call `this`, which will refer to `duck`

Our code for `makeFriend` will look like this:

```
var makeFriend = function(newFriend) {
    this.friends.push(newFriend);
};
```

## Actions to take

Fix the syntax errors in your code (with the help of the Chrome console). Once you see no errors, what else could you improve?

jblneely committed:
revised

jblneely commented:
Thank you for your detailed notes.  

I fixed the syntax errors.  (commas after he hunger level properties in the animal objects in Part 1, a bracket at the end of Part 3, and a comma after the duck name property in Part 5.)  I removed the this from the for loop calling the animal noises.  I also changed the condition of the call from <= to simply < which fixed an error message on that call after the animal noises were returned.  I also:

~added console.log to the hunger function, but it is logging "cannot set property of hunger undefined."  I'm not sure how to fix this.

~made the duck and swan code reciprocal.  Because adding and setting a value to my hunger property in the step above doesn't work, the declination to the hunger property value in the last line of code doesn't act on anything.
